src/remote_gym/remote_environment_management.py

```diff
@@ -85,111 +95,239 @@ def start_as_remote_environment(
+            observation_spec[1], minimum=observation_space_bounds[0], maximum=observation_space_bounds[1]
         )
 
         tensor_spec_utils.set_bounds(
-            self.observation_spec[2],
-            minimum=self.environment.reward_range[0],
-            maximum=self.environment.reward_range[1],
+            observation_spec[2],
+            minimum=env.reward_range[0],
+            maximum=env.reward_range[1],
         )
 
-        if self.rendering_enabled:
-            assert self.environment.render_mode == "rgb_array", (
+        if enable_rendering:
+            assert env.render_mode == "rgb_array", (
                 "Rendering remote environments is only possible if the `render_mode` attribute "
                 "of the passed environment is 'rgb_array'."
             )
-            self.environment.reset()
-            render_shape = self.environment.render().shape
-            self.observation_spec.update(
+            env.reset()
+            render_shape = env.render().shape
+            observation_spec.update(
                 {3: dm_env_rpc_pb2.TensorSpec(name="rendering", shape=render_shape, dtype=dm_env_rpc_pb2.UINT8)}
             )
             tensor_spec_utils.set_bounds(
-                self.observation_spec[3],
+                observation_spec[3],
                 minimum=0,
                 maximum=255,
             )
 
+        # Pass action and obs layout
+        out_queue.put(action_spec)
+        out_queue.put(observation_spec)
+        initialized = True
+
+        while True:
+            action, reset = in_queue.get()
+            if reset is None:
+                break
+
+            if reset:
+                observation, info = env.reset()
+                reward = 0.0
+                terminated = False
+                truncated = False
+            else:
+                observation, reward, terminated, truncated, info = env.step(action)
+
+            rendering = env.render() if enable_rendering else None
+
+            out_queue.put((observation, reward, terminated, truncated, info, rendering))
+
+        env.close()
+    except Exception as e:
+        if not initialized:
+            out_queue.put(None)
+            out_queue.put(None)
+        out_queue.put(e)
+
+        logging.error("Stacktrace: %s", traceback.format_exc())
+
+
+class ProcessedEnv:
+    def __init__(self, args: dict, enable_rendering: bool, env_id: int):
+        self.in_queue = mp.Queue()
+        self.out_queue = mp.Queue()
+        self.should_reset = True
+
+        self.env_id = env_id
+        args["env_id"] = self.env_id
+
+        self.process = mp.Process(target=run_env_loop, args=(args, enable_rendering, self.in_queue, self.out_queue))
+        self.process.start()
+
+        self.action_spec, self.observation_spec = self.out_queue.get(), self.out_queue.get()
+
+        if self.action_spec is None or self.observation_spec is None:
+            raise self.out_queue.get()
```

Comment:
This is very hacky on both sides (here in `ProcessedEnv` as well as `run_env_loop` where you ahndle the Exception by passing None twice and pass the error message afterwards).

Reply:
I will think about a better solution, haven't come up with anything better yet.

Reply:
It might make sense to use a third queue (or maybe something like a bidirectional Pipe even would fit better) to communicate meta information, which both processes always check. 

From Env side it would make sense to pass an initialization dict object in the beginning (with "action_spec" and "observation_spec" as keys) and then check for further objects in case the main process is asks for closing. 

And the loop send back an error message if required.

Reply:
Notice that you then need to handle timeouts etc when fetching the specs, which is what I wanted to avoid.

Reply:
Understood. Maybe then we need to introduce message classes and have different message types.
This way we can still communicate in the same way, but it looks less hacky by assuming that `None` represents something special in certain situations.

Reply:
None does not represent anything special, it is really just the absence of the specs. Since specs are just directories I can pass empty directories instead of that's tidier, but in any case, those values are always ignored.